### ISSUE TITLE
systemd fails to cross-compile

### ISSUE THREAD

**Issue by exarkun** [NixOS/nixpkgs#53207]
## Issue description

When attempting to cross-compile a complete nixos system using nixops, the build fails when it gets to systemd because "meson.build:597:8: ERROR:  Problem encountered: POSIX caps headers not found".  This can also be observed without nixops by just attempting to cross-compile systemd alone. 

### Steps to reproduce

```
nix-build --expr 'with import <nixpkgs> { crossSystem = { system = "armv6l-linux"; }; }; callPackage systemd { }'
```

## Technical details

```
building '/nix/store/5hm4nkdp9h67kdx41hs8v79h8xii47dp-systemd-239-armv6l-unknown-linux-gnueabihf.drv'...
unpacking sources
unpacking source archive /nix/store/f52hdn7hrg3nrgzvlbrswfs6klj4igdh-source
source root is source
patching sources
updateAutotoolsGnuConfigScriptsPhase
configuring
meson flags: --buildtype=release         --libdir=/nix/store/cp86ilwza22rmcfw995n3i2v2iqw7p64-systemd-239-armv6l-unknown-linux-gnueabihf-lib/lib --libexecdir=/nix/store/cp86ilwza22rmcfw995n3i2v2iqw7p64-systemd-239-armv6l-unknown-linux-gnueabihf-lib/libexec         --bindir=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf/bin --sbindir=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf/sbin         --includedir=/nix/store/pd34ndry3rnyi3yr90jz9bcidfj9cq08-systemd-239-armv6l-unknown-linux-gnueabihf-dev/include         --mandir=/nix/store/nkn2yynd8k7iz1h0s27kj82nf6xrj7s3-systemd-239-armv6l-unknown-linux-gnueabihf-man/share/man --infodir=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf/share/info         --localedir=/nix/store/cp86ilwza22rmcfw995n3i2v2iqw7p64-systemd-239-armv6l-unknown-linux-gnueabihf-lib/share/locale         -Dauto_features=disabled         --prefix=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf -Dloadkeys-path=/nix/store/8cyxvc51xvzm6504qh9gvahjh6xp02jk-kbd-2.0.4-armv6l-unknown-linux-gnueabihf/bin/loadkeys -Dsetfont-path=/nix/store/8cyxvc51xvzm6504qh9gvahjh6xp02jk-kbd-2.0.4-armv6l-unknown-linux-gnueabihf/bin/setfont -Dtty-gid=3 -Dlz4=true -Dhostnamed=true -Dnetworkd=true -Dsysusers=false -Dtimedated=true -Dtimesyncd=true -Dfirstboot=false -Dlocaled=true -Dresolve=true -Dsplit-usr=false -Dlibcurl=false -Dlibidn=false -Dlibidn2=true -Dquotacheck=false -Dldconfig=false -Dsmack=true -Dsystem-uid-max=499 -Dsystem-gid-max=499 -Dgnu-efi=false -Defi-libdir=/nix/store/iz576ry8izx3ns2rlv034dqb78x772xx-gnu-efi-3.0.8-armv6l-unknown-linux-gnueabihf/lib -Defi-includedir=/nix/store/iz576ry8izx3ns2rlv034dqb78x772xx-gnu-efi-3.0.8-armv6l-unknown-linux-gnueabihf/include/efi -Defi-ldsdir=/nix/store/iz576ry8izx3ns2rlv034dqb78x772xx-gnu-efi-3.0.8-armv6l-unknown-linux-gnueabihf/lib -Dsysvinit-path= -Dsysvrcnd-path= -Dkill-path=/nix/store/qywhnk0wknfnh3amnzj9b3sacllkj5is-coreutils-8.29-armv6l-unknown-linux-gnueabihf/bin/kill -Dkmod-path=/nix/store/pcphig2piih7d1ra3qb1zl9i9jz6n7gw-kmod-25-armv6l-unknown-linux-gnueabihf/bin/kmod -Dsulogin-path=/nix/store/4g31b0yp5gv2izml1ijyrjcacqiz95q0-util-linux-2.32.1-armv6l-unknown-linux-gnueabihf-bin/bin/sulogin -Dmount-path=/nix/store/4g31b0yp5gv2izml1ijyrjcacqiz95q0-util-linux-2.32.1-armv6l-unknown-linux-gnueabihf-bin/bin/mount -Dumount-path=/nix/store/4g31b0yp5gv2izml1ijyrjcacqiz95q0-util-linux-2.32.1-armv6l-unknown-linux-gnueabihf-bin/bin/umount -Dntp-servers=0.nixos.pool.ntp.org 1.nixos.pool.ntp.org 2.nixos.pool.ntp.org 3.nixos.pool.ntp.org -Ddbuspolicydir=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf/etc/dbus-1/system.d -Ddbussessionservicedir=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf/share/dbus-1/services -Ddbussystemservicedir=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf/share/dbus-1/system-services -Dpamconfdir=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf/etc/pam.d -Drootprefix=/nix/store/0gvrb6b9mvl6q36r9k41y50bkl4914pz-systemd-239-armv6l-unknown-linux-gnueabihf -Drootlibdir=/nix/store/cp86ilwza22rmcfw995n3i2v2iqw7p64-systemd-239-armv6l-unknown-linux-gnueabihf-lib/lib -Dpkgconfiglibdir=/nix/store/pd34ndry3rnyi3yr90jz9bcidfj9cq08-systemd-239-armv6l-unknown-linux-gnueabihf-dev/lib/pkgconfig -Dpkgconfigdatadir=/nix/store/pd34ndry3rnyi3yr90jz9bcidfj9cq08-systemd-239-armv6l-unknown-linux-gnueabihf-dev/share/pkgconfig
The Meson build system
Version: 0.46.1
Source dir: /build/source
Build dir: /build/source/build
Build type: native build
WARNING: Unknown command line options: "auto_features"
This will become a hard error in a future Meson release.
Project name: systemd
Native C compiler: /nix/store/iw94llkj05wgaz268mlzvgx8jkbi1ss0-gcc-wrapper-7.3.0/bin/cc (gcc 7.3.0 "gcc (GCC) 7.3.0")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Program tools/meson-check-compilation.sh found: YES (/build/source/tools/meson-check-compilation.sh)
Program tools/meson-build.sh found: YES (/build/source/tools/meson-build.sh)
Program c++ found: NO
Compiler for C supports arguments -Wextra: YES
Compiler for C supports arguments -Werror=undef: YES
Compiler for C supports arguments -Wlogical-op: YES
Compiler for C supports arguments -Wmissing-include-dirs: YES
Compiler for C supports arguments -Wold-style-definition: YES
Compiler for C supports arguments -Wpointer-arith: YES
Compiler for C supports arguments -Winit-self: YES
Compiler for C supports arguments -Wfloat-equal: YES
Compiler for C supports arguments -Wsuggest-attribute=noreturn: YES
Compiler for C supports arguments -Werror=missing-prototypes: YES
Compiler for C supports arguments -Werror=implicit-function-declaration: YES
Compiler for C supports arguments -Werror=missing-declarations: YES
Compiler for C supports arguments -Werror=return-type: YES
Compiler for C supports arguments -Werror=incompatible-pointer-types: YES
Compiler for C supports arguments -Werror=format=2: YES
Compiler for C supports arguments -Wstrict-prototypes: YES
Compiler for C supports arguments -Wredundant-decls: YES
Compiler for C supports arguments -Wmissing-noreturn: YES
Compiler for C supports arguments -Wimplicit-fallthrough=5: YES
Compiler for C supports arguments -Wshadow: YES
Compiler for C supports arguments -Wendif-labels: YES
Compiler for C supports arguments -Wstrict-aliasing=2: YES
Compiler for C supports arguments -Wwrite-strings: YES
Compiler for C supports arguments -Werror=overflow: YES
Compiler for C supports arguments -Werror=shift-count-overflow: YES
Compiler for C supports arguments -Werror=shift-overflow=2: YES
Compiler for C supports arguments -Wdate-time: YES
Compiler for C supports arguments -Wnested-externs: YES
Compiler for C supports arguments -ffast-math: YES
Compiler for C supports arguments -fno-common: YES
Compiler for C supports arguments -fdiagnostics-show-option: YES
Compiler for C supports arguments -fno-strict-aliasing: YES
Compiler for C supports arguments -fvisibility=hidden: YES
Compiler for C supports arguments -fstack-protector: YES
Compiler for C supports arguments -fstack-protector-strong: YES
Compiler for C supports arguments --param=ssp-buffer-size=4: YES
Compiler for C supports arguments -fPIE: YES
Compiler for C supports arguments -ffunction-sections: YES
Compiler for C supports arguments -fdata-sections: YES
Compiler for C supports arguments -Wunused-parameter: YES
Compiler for C supports arguments -Wmissing-field-initializers: YES
Compiler for C supports arguments -Wunused-result: YES
Compiler for C supports arguments -Wformat-signedness: YES
Compiler for C supports arguments -Werror=nonnull: YES
Checking if "-Werror=shadow with local shadowing" compiles: YES
Message: Linking with -Wl,-z,relro supported: yes
Message: Linking with -Wl,-z,now supported: yes
Message: Linking with -pie supported: yes
Message: Linking with -Wl,--gc-sections supported: yes
Checking for size of "pid_t": 4
Checking for size of "uid_t": 4
Checking for size of "gid_t": 4
Checking for size of "dev_t": 8
Checking for size of "ino_t": 8
Checking for size of "time_t": 8
Checking for size of "rlim_t": 8
Checking for size of "char16_t": 2
Checking for size of "char32_t": 4
Checking for size of "key_serial_t": -1
Checking for size of "struct ethtool_link_settings": 48
Checking for size of "struct fib_rule_uid_range": 8
Checking for size of "struct statx": 256
Header <linux/if_link.h> has symbol "IFLA_INET6_ADDR_GEN_MODE": YES
Header <linux/if_link.h> has symbol "IN6_ADDR_GEN_MODE_STABLE_PRIVACY": YES
Header <linux/if_link.h> has symbol "IFLA_VRF_TABLE": YES
Header <linux/if_link.h> has symbol "IFLA_MACVLAN_FLAGS": YES
Header <linux/if_link.h> has symbol "IFLA_IPVLAN_FLAGS": YES
Header <linux/if_link.h> has symbol "IFLA_PHYS_PORT_ID": YES
Header <linux/if_link.h> has symbol "IFLA_BOND_AD_INFO": YES
Header <linux/if_link.h> has symbol "IFLA_VLAN_PROTOCOL": YES
Header <linux/if_link.h> has symbol "IFLA_VXLAN_REMCSUM_NOPARTIAL": YES
Header <linux/if_link.h> has symbol "IFLA_VXLAN_GPE": YES
Header <linux/if_link.h> has symbol "IFLA_GENEVE_LABEL": YES
Header <linux/if_tunnel.h> has symbol "IFLA_VTI_REMOTE": YES
Header <linux/if_tunnel.h> has symbol "IFLA_IPTUN_ENCAP_DPORT": YES
Header <linux/if_tunnel.h> has symbol "IFLA_GRE_ENCAP_DPORT": YES
Header <linux/if_bridge.h> has symbol "IFLA_BRIDGE_VLAN_INFO": YES
Header <linux/if_link.h> has symbol "IFLA_BRPORT_PROXYARP": YES
Header <linux/if_link.h> has symbol "IFLA_BRPORT_LEARNING_SYNC": YES
Header <linux/if_link.h> has symbol "IFLA_BR_VLAN_DEFAULT_PVID": YES
Header <linux/if_link.h> has symbol "IPVLAN_F_PRIVATE": YES
Header <linux/neighbour.h> has symbol "NDA_IFINDEX": YES
Header <linux/if_addr.h> has symbol "IFA_FLAGS": YES
Header <linux/fib_rules.h> has symbol "FRA_UID_RANGE": YES
Header <linux/loop.h> has symbol "LO_FLAGS_PARTSCAN": YES
Header <linux/can/vxcan.h> has symbol "VXCAN_INFO_PEER": YES
Checking for function "secure_getenv": YES
Checking for function "__secure_getenv": NO
Checking for function "memfd_create": YES
Checking for function "gettid": NO
Checking for function "pivot_root": NO
Checking for function "name_to_handle_at": YES
Checking for function "setns": YES
Checking for function "renameat2": NO
Checking for function "kcmp": NO
Checking for function "keyctl": NO
Checking for function "copy_file_range": YES
Checking for function "bpf": NO
Checking for function "statx": NO
Checking for function "explicit_bzero": YES
Checking for function "reallocarray": YES
Checking for function "getrandom": YES
Program sed found: YES (/nix/store/ny5p32137wfyzdm485xfdck21w1gyl3g-gnused-4.5/bin/sed)
Program awk found: YES (/nix/store/zdadpaj5z8k9ifkgsc16yvfjn4wdv0r2-gawk-4.2.1/bin/awk)
Program m4 found: YES (/nix/store/fr2a9kwbcvz4zn505lc3l1v2i61f2cph-gnum4-1.4.18/bin/m4)
Program stat found: YES (/nix/store/wm8va53fh5158ipi0ic9gir64hrvqv1z-coreutils-8.29/bin/stat)
Program git found: NO
Program env found: YES (/nix/store/wm8va53fh5158ipi0ic9gir64hrvqv1z-coreutils-8.29/bin/env)
Program perl found: YES (/nix/store/q9ajsngf2w3r10dg7b1148wzq1avp4gj-perl-5.28.0/bin/perl)
Program test/test-efi-create-disk.sh found: YES (/build/source/test/test-efi-create-disk.sh)
Program quotaon found: NO
Program /usr/sbin/quotaon found: NO
Program /sbin/quotaon found: NO
Program quotacheck found: NO
Program /usr/sbin/quotacheck found: NO
Program /sbin/quotacheck found: NO
Message: Using /nix/store/pcphig2piih7d1ra3qb1zl9i9jz6n7gw-kmod-25-armv6l-unknown-linux-gnueabihf/bin/kmod for kmod
Program kexec found: NO
Program /usr/sbin/kexec found: NO
Program /sbin/kexec found: NO
Message: Using /nix/store/4g31b0yp5gv2izml1ijyrjcacqiz95q0-util-linux-2.32.1-armv6l-unknown-linux-gnueabihf-bin/bin/sulogin for sulogin
Message: Using /nix/store/4g31b0yp5gv2izml1ijyrjcacqiz95q0-util-linux-2.32.1-armv6l-unknown-linux-gnueabihf-bin/bin/mount for mount
Message: Using /nix/store/4g31b0yp5gv2izml1ijyrjcacqiz95q0-util-linux-2.32.1-armv6l-unknown-linux-gnueabihf-bin/bin/umount for umount
Message: Using /nix/store/8cyxvc51xvzm6504qh9gvahjh6xp02jk-kbd-2.0.4-armv6l-unknown-linux-gnueabihf/bin/loadkeys for loadkeys
Message: Using /nix/store/8cyxvc51xvzm6504qh9gvahjh6xp02jk-kbd-2.0.4-armv6l-unknown-linux-gnueabihf/bin/setfont for setfont
Program gperf found: YES (/nix/store/9p6mxkadz27vv9afks3y994pwjsc9vn5-gperf-3.1/bin/gperf)
Message: gperf len type is size_t
Has header "sys/capability.h": NO

meson.build:597:8: ERROR:  Problem encountered: POSIX caps headers not found

A full log can be found at /build/source/build/meson-logs/meson-log.txt
builder for '/nix/store/5hm4nkdp9h67kdx41hs8v79h8xii47dp-systemd-239-armv6l-unknown-linux-gnueabihf.drv' failed with exit code 1
cannot build derivation '/nix/store/pr24b944wzkqir7pqsp7757d6dvzj74k-nixos-system-rpi1-18.09.910.c15e342304a-armv6l-unknown-linux-gnueabihf.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/2lckxiq5f4li288jk0s31wkb379m5ggf-nixops-machines.drv': 1 dependencies couldn't be built
error: build of '/nix/store/2lckxiq5f4li288jk0s31wkb379m5ggf-nixops-machines.drv' failed
Traceback (most recent call last):
  File "/nix/store/0pd48alphnpv3fmdbczv61s78z4ld3hx-nixops-1.6/bin/..nixops-wrapped-wrapped", line 985, in <module>
    args.op()
  File "/nix/store/0pd48alphnpv3fmdbczv61s78z4ld3hx-nixops-1.6/bin/..nixops-wrapped-wrapped", line 407, in op_deploy
    max_concurrent_activate=args.max_concurrent_activate)
  File "/nix/store/0pd48alphnpv3fmdbczv61s78z4ld3hx-nixops-1.6/lib/python2.7/site-packages/nixops/deployment.py", line 1051, in deploy
    self.run_with_notify('deploy', lambda: self._deploy(**kwargs))
  File "/nix/store/0pd48alphnpv3fmdbczv61s78z4ld3hx-nixops-1.6/lib/python2.7/site-packages/nixops/deployment.py", line 1040, in run_with_notify
    f()
  File "/nix/store/0pd48alphnpv3fmdbczv61s78z4ld3hx-nixops-1.6/lib/python2.7/site-packages/nixops/deployment.py", line 1051, in <lambda>
    self.run_with_notify('deploy', lambda: self._deploy(**kwargs))
  File "/nix/store/0pd48alphnpv3fmdbczv61s78z4ld3hx-nixops-1.6/lib/python2.7/site-packages/nixops/deployment.py", line 991, in _deploy
    self.configs_path = self.build_configs(dry_run=dry_run, repair=repair, include=include, exclude=exclude)
  File "/nix/store/0pd48alphnpv3fmdbczv61s78z4ld3hx-nixops-1.6/lib/python2.7/site-packages/nixops/deployment.py", line 659, in build_configs
    raise Exception("unable to build all machine configurations")
Exception: unable to build all machine configurations
```

It appears that meson has not been told to attempt to cross-compile.  Instead, meson thinks it should do a native build (which is maybe relevant, I don't know).

```
[root@ip-172-30-5-31:~]# nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.76, NixOS, 18.09.910.c15e342304a (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by exarkun** [comment]
Looking more closely, I see a few things of note.

First, pkgs/development/tools/build-managers/meson/default.nix compares stdenv.buildPlatform to stdenv.hostPlatform to determine if cross compilation is happening.  However, meson is a build tool.  It drives compilation *on* the host for a *target*.  So shouldn't there be a comparison of hostPlatform against targetPlatform to determine if cross compilation is happening?

Second, when pkgs/development/tools/build-managers/meson/default.nix writes out the cross-file.conf (as it *always* does - it just doesn't always use it) it finds "" as all of the targetPrefixes so the commands it writes out are like "cc", "ar", etc.  I don't understand why the targetPrefix is not set correctly.

**Comment by FRidh** [comment]
I did not test on master or any of the channels, but with https://github.com/NixOS/nixpkgs/pull/53123 I was able to cross-compile `systemd`:
```
$ nix-build -A pkgsCross.raspberryPi.systemd
/nix/store/m69vz0zsx9yfqv94l7k8zgsfdwszadc6-systemd-239-armv6l-unknown-linux-gnueabihf
```
It also succeeded with `arm64-multiplatform`. Note there are still some remaining references to the build packages.

**Comment by exarkun** [comment]
Thanks.  I tried against with staging where #53123 has been merged (not sure if #53123 fixed it or if some other merge on staging did) and systemd did cross-compile.  Thanks.
